Source/JavaScriptCore/fuzzilli/Fuzzilli.cpp

```diff
@@ -0,0 +1,171 @@
+/*
+ *  Copyright (C) 2023 Apple Inc. All rights reserved.
+ *
+ *  This library is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Library General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2 of the License, or (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Library General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Library General Public License
+ *  along with this library; see the file COPYING.LIB.  If not, write to
+ *  the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ *  Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "config.h"
+#include "Fuzzilli.h"
+
+#include <mutex>
+#include <wtf/Assertions.h>
+
+#if ENABLE(FUZZILLI)
+
+#define REPRL_CRFD 100
+#define REPRL_CWFD 101
+#define REPRL_DRFD 102
+#define REPRL_DWFD 103
+#define REPRL_MAX_DATA_SIZE (16 * 1024 * 1024)
+
+#define SHM_SIZE 0x100000
+#define MAX_EDGES ((SHM_SIZE - 4) * 8)
+
+#define WRITE_TO_FUZZILLI(data_, size_) RELEASE_ASSERT(write(REPRL_CWFD, data_, size_) == static_cast<ssize_t>(size_))
+#define READ_FROM_FUZZILLI(data_, size_) RELEASE_ASSERT(read(REPRL_CRFD, data_, size_) == static_cast<ssize_t>(size_))
+
+struct Fuzzilli::SharedData* Fuzzilli::sharedData { nullptr };
+
+uint32_t* Fuzzilli::edgesStart { nullptr };
+uint32_t* Fuzzilli::edgesStop { nullptr };
+
+char* Fuzzilli::reprlInputData { nullptr };
+size_t Fuzzilli::numPendingRejectedPromises { 0 };
+
+void Fuzzilli::resetCoverageEdges()
+{
+    uint64_t n = 0;
+    for (uint32_t* edge = edgesStart; edge < edgesStop && n < MAX_EDGES; edge++)
+        *edge = ++n;
+}
+
+FILE* Fuzzilli::logFile()
+{
+    static FILE* result;
+    static std::once_flag flag;
+    std::call_once(
+        flag,
+        [] () {
+            result = fdopen(REPRL_DWFD, "w");
+            if (!result) {
+                result = stdout;
+
+                fprintf(stderr, "Fuzzer output channel not available, printing to stdout instead\n");
```

Comment:
Let's use `dataLogLn("Fuzzer output channel not available, printing to stdout instead");`